File: backend/src/main/java/ru/home/shop/domain/Present.java
package ru.home.shop.domain;

import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.Type;

import javax.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.*;

@Getter
@Setter
@Entity
public class Present {

    @Id
    @Type(type = "uuid-char")
    private UUID id;

    private String name;
    private BigDecimal price;
    private LocalDateTime date;

    @ElementCollection
    @CollectionTable(name = "present_item", joinColumns = @JoinColumn(name = "present_id"))
    private List<Item> items = new ArrayList<>();

    public BigDecimal computeCost() {
        return items.stream()
                .map(item -> item.getCandy().getPrice().multiply(BigDecimal.valueOf(item.getCount())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
}

File: backend/src/main/kotlin/io/github/artemy/osipov/shop/service/present/Present.kt
package io.github.artemy.osipov.shop.service.present

import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Table
import java.util.UUID
import java.math.BigDecimal
import java.time.LocalDateTime

data class Present(
    @Id
    val id: UUID,
    var name: String,
    var price: BigDecimal,
    var createDate: LocalDateTime,
    val items: MutableList<Item> = mutableListOf()
) {
    fun addItem(candyId: UUID, count: Int) {
        items.add(
            Item(candyId, count)
        )
    }

    @Table("present_items")
    data class Item(
        val candyId: UUID,
        var count: Int
    )
}
